src/multinomials.jl

```diff
@@ -0,0 +1,55 @@
+# Multinomial theorem
+# https://en.wikipedia.org/wiki/Multinomial_theorem
+
+export multiexponents
+
+struct MultiExponents{T}
+    c::Combinations{T}
+    nterms::Int
+end
+
+start(m::MultiExponents) = start(m.c)
+
+function next(m::MultiExponents, s)
+    stars, ss = next(m.c, s)
+
+    # stars minus their consecutive
+    # position becomes their index
+    idx = stars - [0:length(stars)-1;]
+
+    result = zeros(Int, m.nterms)
+    for i in idx
+      result[i] += 1
+    end
```

Comment:
I think this whole function could be
```julia
function next(m::MultiExponents, s)
    stars, ss = next(m.c, s)
    result = zeros(Int, m.nterms)
    result[stars .- 0:length(stars)-1] .+= 1
    result, ss
end
```
This _should_ be fusing on 0.6 and above but I'm not certain.

Reply:
Unfortunately they are not equivalent, the `idx` variable can contain repeated indexes, the fusion won't work. I thought about it to save some temporary variables.

Reply:
`a[foo] .+= bar` doesn't fuse `foo` with `bar`.

Reply:
To do it with one fewer allocation you would need something like
```jl
result = zeros(Int, m.nterms)
for (i,s) in enumerate(stars)
    result[s-i+1] += 1
end
```
Even then, `next(m.c, s)` is still allocating... could there be a `next!(m.c, s)` method that overwrites a preallocated `s` in-place?

Reply:
Thank you @stevengj , very good comments, I will try to address them tomorrow.